studies/migrations/0043_add_video_constructs.py

```diff
@@ -0,0 +1,205 @@
+    but it's quite possible we'll have the need for dynamic upsertion later.
+    """
+
+    seen_ids = set()
+    video_objects = []
+    Video = video_model
+
+    # Using a constructive approach here, but with an ancillary seen_ids list b/c Django models without
+    # primary keys are unhashable for some dumb reason (even though they have unique fields...)
+    for frame_id, event_data in response.exp_data.items():
+        if event_data.get("videoList", None) and event_data.get("videoId", None):
+            # We've officially captured video here!
+            events = event_data.get("eventTimings", [])
+            for event in events:
+                video_id = event["videoId"]
+                pipe_name = event["pipeId"]  # what we call "ID" they call "name"
+                if video_id not in seen_ids and pipe_name and event["streamTime"] > 0:
+                    # Try looking for the regular ID first.
+                    file_obj = S3_RESOURCE.Object(
+                        settings.BUCKET_NAME, f"{video_id}.mp4"
+                    )
+                    try:
+                        s3_response = file_obj.get()
+                    except ClientError:
+                        try:  # If that doesn't work, use the pipe name.
+                            file_obj = S3_RESOURCE.Object(
+                                settings.BUCKET_NAME, f"{pipe_name}.mp4"
+                            )
+                            s3_response = file_obj.get()
+                        except ClientError:
+                            logger.warning(
+                                f"could not find {video_id} or {pipe_name} in S3!"
+                            )
+                            continue
+                    # Read first 32 bytes from streaming body (file header) to get actual filetype.
+                    streaming_body = s3_response["Body"]
+                    file_header_buffer: bytes = streaming_body.read(32)
+                    file_info = fleep.get(file_header_buffer)
+                    streaming_body.close()
+
+                    video_objects.append(
+                        Video(
+                            pipe_name=pipe_name,
+                            created_at=date_parser.parse(event["timestamp"]),
+                            date_modified=s3_response["LastModified"],
+                            #  Can't get the *actual* pipe id property, it's in the webhook payload...
+                            frame_id=frame_id,
+                            full_name=f"{video_id}.{file_info.extension[0]}",
+                            study=response.study,
+                            response=response,
+                            is_consent_footage=frame_id in VALID_CONSENT_FRAMES,
+                        )
+                    )
+                    seen_ids.add(video_id)
+
+    return Video.objects.bulk_create(video_objects)
+
+
+class Migration(migrations.Migration):
```

Comment:
Is the idea that once this is deployed to production we’ll just have researchers redo consent coding via the UI?  Pros and cons of creating ‘accepted’ rulings for existing records with ‘comments’ field set to something like ‘automatically accepted and may not be verified; recorded before consent manager deployed’?

Reply:
We absolutely could create ConsentRulings in bulk with that comment, we'd just need to acquire/create some sort of fake user/bot that would serve as the arbiter for each

Reply:
alternatively, we could use your account